indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskClient.java

```diff
@@ -0,0 +1,370 @@
+
+    try {
+      final FullResponseHolder response = submitRequestWithEmptyContent(id, HttpMethod.GET, "offsets/end", null, true);
+      return deserialize(response.getContent(), constructPartitionOffsetMapType(Map.class));
+    }
+    catch (NoTaskLocationException e) {
+      return ImmutableMap.of();
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public boolean setEndOffsets(
+      final String id,
+      final Map<PartitionType, SequenceType> endOffsets,
+      final boolean finalize
+  ) throws IOException
+  {
+    log.debug("SetEndOffsets task[%s] endOffsets[%s] finalize[%s]", id, endOffsets, finalize);
+
+    try {
+      final FullResponseHolder response = submitJsonRequest(
+          id,
+          HttpMethod.POST,
+          "offsets/end",
+          StringUtils.format("finish=%s", finalize),
+          serialize(endOffsets),
+          true
+      );
+      return isSuccess(response);
+    }
+    catch (NoTaskLocationException e) {
+      return false;
+    }
+  }
+
+  public ListenableFuture<Boolean> stopAsync(final String id, final boolean publish)
+  {
+    return doAsync(() -> stop(id, publish));
+  }
+
+
+  public ListenableFuture<Boolean> resumeAsync(final String id)
+  {
+    return doAsync(() -> resume(id));
+  }
+
+
+  public ListenableFuture<DateTime> getStartTimeAsync(final String id)
+  {
+    return doAsync(() -> getStartTime(id));
+  }
+
+
+  public ListenableFuture<Map<PartitionType, SequenceType>> pauseAsync(final String id)
+  {
+    return doAsync(() -> pause(id));
+  }
+
+  public ListenableFuture<Boolean> setEndOffsetsAsync(
+      final String id,
+      final Map<PartitionType, SequenceType> endOffsets,
+      final boolean finalize
+  )
+  {
+    return doAsync(() -> setEndOffsets(id, endOffsets, finalize));
+  }
+
+  public ListenableFuture<Map<PartitionType, SequenceType>> getCurrentOffsetsAsync(final String id, final boolean retry)
+  {
+    return doAsync(() -> getCurrentOffsets(id, retry));
+  }
+
+  public ListenableFuture<Map<PartitionType, SequenceType>> getEndOffsetsAsync(final String id)
+  {
+    return doAsync(() -> getEndOffsets(id));
+  }
+
+
+  public ListenableFuture<Map<String, Object>> getMovingAveragesAsync(final String id)
+  {
+    return doAsync(() -> getMovingAverages(id));
+  }
+
+
+  public ListenableFuture<SeekableStreamIndexTaskRunner.Status> getStatusAsync(final String id)
+  {
+    return doAsync(() -> getStatus(id));
+  }
+
+  private JavaType constructCheckpointMapType()
+  {
+    ObjectMapper mapper = new ObjectMapper();
+    return mapper.getTypeFactory()
+                 .constructMapType(
+                     TreeMap.class,
+                     mapper.getTypeFactory().constructType(Integer.class),
+                     constructPartitionOffsetMapType(TreeMap.class)
+                 );
+  }
+
+  protected abstract JavaType constructPartitionOffsetMapType(Class<? extends Map> mapType);
```

Comment:
Since `objectMapper` is private in `IndexTaskClient` and not accessible from the child classes, you added a new objectMapper [here](https://github.com/apache/incubator-druid/pull/6431/files#diff-edc0abdecae99c05a70ae5436d115827R34). However, I don't think this is a good idea.

Instead of this method, I would suggest to add two methods to get the `Class` of `PartitionType` and `SequenceType` and add `deserializeMap` method to `IndexTaskClient` like below.

in SeekableStreamIndexTaskClient,

```
protected abstract Class<PartitionType> getPartitionType();
protected abstract Class<SequenceType> getSequenceType();
```

In IndexTaskClient,
```
  protected <T> T deserializeMap(String content, Class<? extends Map> mapClass, Class<?> keyClass, Class<?> valueClass) throws IOException
  {
    return deserialize(content, objectMapper.getTypeFactory().constructMapType(mapClass, keyClass, valueClass));
  }
```

Reply:
added